File: public-api/examples/typescript/streaming-search.ts
/**
 * Streaming Search Example (TypeScript)
 * 
 * This example demonstrates how to use Server-Sent Events (SSE) streaming
 * to receive search results in real-time with TypeScript.
 */

interface StreamingProduct {
  product_id: string;
  impression_id: string;
  title: string;
  price: string;
  relevance_score: number;
}

interface StreamingData {
  results?: StreamingProduct[];
  metadata?: {
    total_matches?: number;
  };
}

/**
 * Perform a streaming search using Server-Sent Events.
 * 
 * @param query - The search query
 * @param apiKey - Your Ad-Tokens API key
 */
async function streamSearch(query: string, apiKey: string): Promise<void> {
  try {
    const response = await fetch('https://api.ad-tokens.com/search', {
      method: 'POST',
      headers: {
        'x-api-key': apiKey,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: query,
        limit: 5,
        stream: true, // Enable streaming
      }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(
        `HTTP error! status: ${response.status}, message: ${error.message}`
      );
    }

    if (!response.body) {
      throw new Error('Response body is null');
    }

    console.log(`🔍 Streaming search for: ${query}\n`);
    console.log('📡 Waiting for results...\n');

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let productsReceived = 0;

    while (true) {
      const { done, value } = await reader.read();

      if (done) {
        break;
      }

      // Decode chunk and add to buffer
      buffer += decoder.decode(value, { stream: true });

      // Process complete lines
      const lines = buffer.split('\n');
      buffer = lines.pop() || ''; // Keep incomplete line in buffer

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;

        // SSE format: "event: result\ndata: {...}\n\n"
        if (trimmed.startsWith('event: ')) {
          const eventType = trimmed.slice(7);
          // Handle event type if needed
          continue;
        }

        if (trimmed.startsWith('data: ')) {
          try {
            const data: StreamingData = JSON.parse(trimmed.slice(6));

            // Handle different event types
            if (data.results) {
              for (const product of data.results) {
                productsReceived++;
                console.log(
                  `✅ Product ${productsReceived}: ${product.title}`
                );
                console.log(`   Price: ${product.price}`);
                console.log(
                  `   Relevance: ${(product.relevance_score * 100).toFixed(2)}%\n`
                );
              }
            }

            if (data.metadata?.total_matches) {
              console.log(`📊 Total matches: ${data.metadata.total_matches}`);
            }
          } catch (e) {
            if (e instanceof Error) {
              console.warn(`⚠️  Failed to parse JSON: ${e.message}`);
            }
            console.warn(`   Line: ${trimmed}`);
          }
        }
      }
    }

    console.log(`\n✨ Received ${productsReceived} products via streaming`);
  } catch (error) {
    if (error instanceof Error) {
      console.error(`❌ Error: ${error.message}`);
    } else {
      console.error(`❌ Error: ${error}`);
    }
  }
}

/**
 * Main function to demonstrate streaming search.
 */
async function main(): Promise<void> {
  const apiKey = process.env.AD_TOKENS_API_KEY || 'your-api-key-here';

  if (apiKey === 'your-api-key-here') {
    console.error('⚠️  Please set AD_TOKENS_API_KEY environment variable');
    return;
  }

  const query = 'mechanical keyboard for programming';
  await streamSearch(query, apiKey);
}

// Run if this is the main module
if (require.main === module) {
  main();
}

// Export for use in other modules
export { streamSearch };

File: public-api/examples/python/streaming-search.py
"""
Streaming Search Example

This example demonstrates how to use Server-Sent Events (SSE) streaming
to receive search results in real-time.
"""

import asyncio
import json
import os
import httpx


async def stream_search(query: str, api_key: str):
    """
    Perform a streaming search using Server-Sent Events.
    
    Args:
        query: The search query
        api_key: Your Ad-Tokens API key
    """
    async with httpx.AsyncClient() as client:
        try:
            async with client.stream(
                "POST",
                "https://api.ad-tokens.com/search",
                headers={
                    "x-api-key": api_key,
                    "Content-Type": "application/json",
                },
                json={
                    "query": query,
                    "limit": 5,
                    "stream": True,  # Enable streaming
                },
                timeout=30.0,
            ) as response:
                response.raise_for_status()
                
                print(f"🔍 Streaming search for: {query}\n")
                print("📡 Waiting for results...\n")
                
                buffer = ""
                products_received = 0
                
                async for chunk in response.aiter_text():
                    buffer += chunk
                    
                    # Process complete lines
                    while "\n" in buffer:
                        line, buffer = buffer.split("\n", 1)
                        line = line.strip()
                        
                        if not line:
                            continue
                        
                        # SSE format: "event: result\ndata: {...}\n\n"
                        if line.startswith("event: "):
                            event_type = line[7:]
                            continue
                        
                        if line.startswith("data: "):
                            try:
                                data = json.loads(line[6:])
                                
                                # Handle different event types
                                if "results" in data:
                                    for product in data["results"]:
                                        products_received += 1
                                        print(f"✅ Product {products_received}: {product['title']}")
                                        print(f"   Price: {product['price']}")
                                        print(f"   Relevance: {product['relevance_score']:.2%}\n")
                                
                                if "metadata" in data:
                                    metadata = data["metadata"]
                                    if "total_matches" in metadata:
                                        print(f"📊 Total matches: {metadata['total_matches']}")
                                        
                            except json.JSONDecodeError as e:
                                print(f"⚠️  Failed to parse JSON: {e}")
                                print(f"   Line: {line}")
                
                print(f"\n✨ Received {products_received} products via streaming")
                
        except httpx.HTTPStatusError as e:
            print(f"❌ HTTP error: {e.response.status_code}")
            print(f"   Error: {e.response.text}")
        except httpx.RequestError as e:
            print(f"❌ Request error: {e}")


async def main():
    api_key = os.getenv("AD_TOKENS_API_KEY", "your-api-key-here")
    
    if api_key == "your-api-key-here":
        print("⚠️  Please set AD_TOKENS_API_KEY environment variable")
        return
    
    query = "mechanical keyboard for programming"
    await stream_search(query, api_key)


if __name__ == "__main__":
    asyncio.run(main())

